### PR TITLE
ENH: Add/reorganise metrics and correctly normalise confusion matrix

### PR DESCRIPTION
**Pull request by harshita-s**
In this PR:
- Average precision metric is added to multi-class (n_classes > 1) case.
- AUROC is modified such that `num_classes=None` for binary case, as prescribed in Pytorch documentation https://torchmetrics.readthedocs.io/en/stable/classification/auroc.html.
- Parameter `num_classes` is not explicitly given in Specificity, similar to the other binary metrics.
- Hardcoded `threshold=0.5` is removed from binary metrics (since this is default value).
- Confusion matrix is normalized on true values.
- Metrics are re-organized for readability.

### CODE REVIEW

File: hi-ml-cpath/src/health_cpath/models/deepmil.py [hunk 1]
```diff
@@ -190,29 +190,32 @@ def get_bag_label(labels: Tensor) -> Tensor:
...
-                                  MetricsKey.F1: F1(threshold=threshold),
-                                  MetricsKey.CONF_MATRIX: ConfusionMatrix(num_classes=2, threshold=threshold),
-                                  # Average precision is a measure of area under the PR curve
-                                  # https://sanchom.wordpress.com/tag/average-precision/
-                                  MetricsKey.AVERAGE_PRECISION: AveragePrecision(),
-                                  MetricsKey.COHENKAPPA: CohenKappa(num_classes=2, weights='quadratic',
-                                                                    threshold=threshold),
-                                  MetricsKey.SPECIFICITY: Specificity(num_classes=self.n_classes, threshold=threshold)})
+            return nn.ModuleDict({
+                MetricsKey.ACC: Accuracy(threshold=threshold),
```
Comment (vale-salvatelli) [inline hi-ml-cpath/src/health_cpath/models/deepmil.py:208]: I am still confused about what would happen in n_classes = 2 vs n_classes = 1. I think our pipeline should give equivalent results because in both cases we would be working with a binary label while here if n_classes=1 we take a threshold of 0.5, while if n_classes=2 we pass num_classes=2

Comment (harshita-s) [reply hi-ml-cpath/src/health_cpath/models/deepmil.py:208]: I am investigating this aspect.

Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/deepmil.py:208]: Note that threshold is always 0.5 by default in all metrics, not sure if we should override it for binary case given that it will be set to 0.5 by default in multi class as well. (Sorry for jumping in I was just curious)

Comment (harshita-s) [reply hi-ml-cpath/src/health_cpath/models/deepmil.py:208]: Yes, I am aware of the default values. I am not sure why these were hardcoded in the first instance, perhaps @vale-salvatelli could clarify?

Comment (harshita-s) [reply hi-ml-cpath/src/health_cpath/models/deepmil.py]: I have now removed the hardcoded threshold for binary case

File: hi-ml-cpath/src/health_cpath/models/deepmil.py [hunk 2]
```diff
@@ -190,29 +190,31 @@ def get_bag_label(labels: Tensor) -> Tensor:
...
-                                  # https://sanchom.wordpress.com/tag/average-precision/
-                                  MetricsKey.AVERAGE_PRECISION: AveragePrecision(),
-                                  MetricsKey.COHENKAPPA: CohenKappa(num_classes=2, weights='quadratic',
-                                                                    threshold=threshold),
-                                  MetricsKey.SPECIFICITY: Specificity(num_classes=self.n_classes, threshold=threshold)})
+            return nn.ModuleDict({
+                MetricsKey.ACC: Accuracy(),
+                MetricsKey.AUROC: AUROC(num_classes=None),
+                # Average precision is a measure of area under the PR curve
+                MetricsKey.AVERAGE_PRECISION: AveragePrecision(),
```
Comment (vale-salvatelli) [inline hi-ml-cpath/src/health_cpath/models/deepmil.py:210]: why we don't pass self.n_classes in this case?

**Review by vale-salvatelli — APPROVED** [review]
I think this PR doesn't solve the inconsistency between num_classes 1 and num_classes 2 but it contains some fixes we should merge soon - OK to merge for me. 

I would probably change the title to say it also fixes the confusion matrix

File: hi-ml-cpath/src/health_cpath/models/deepmil.py (hunk 2 above)
Comment (harshita-s) [reply hi-ml-cpath/src/health_cpath/models/deepmil.py:210]: I followed the docs https://torchmetrics.readthedocs.io/en/stable/classification/accuracy.html to determine the parameters for each case.

**Comment by harshita-s** [comment]
> I think this PR doesn't solve the inconsistency between num_classes 1 and num_classes 2 but it contains some fixes we should merge soon - OK to merge for me.
> 
> I would probably change the title to say it also fixes the confusion matrix

The class discrepency will be handled in a different PR or summarized on OneNote.
